### DISCUSSION TITLE
What is keepAlive option in http.createServer actual for?

### DISCUSSION THREAD

**Discussion by linhx** [nodejs/help#4748] [Q&A]
Hi, I'm using keepAlive for http connection.
But I'm confused on the option keepAlive in http.createServer.

> keepAlive [boolean](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Boolean_type) If set to true, it enables keep-alive functionality on the socket immediately after a new incoming connection is received, similarly on what is done in [socket.setKeepAlive([enable][, initialDelay])][socket.setKeepAlive(enable, initialDelay)]. Default: false.

I understand if it's false, it means server will not keep connection alive. But I've tried to false, the keep-alive still works if I set the keepAlive of Agent of client to true.
please help me to understand what the `keepAlive` really means?

My example:

env: Ubuntu 22, Node 18.16.1

```js
// server.js
var http = require('http');

let prevSocket;
http
  .createServer({
    keepAlive: false,
  }, async function (req, res) {
    const isPrev = req.socket === prevSocket;
    console.log('is prev', isPrev)
    if (!isPrev) {
      req.socket.on('close', () => {
        console.log('close', req.url);
      });
    }
    prevSocket = req.socket;
    res.write('Hello World!');
    res.end();
  })
  .listen(3000);
```

```js
// client.js
const fetch = require('node-fetch');
const http = require('http');

const httpAgent = new http.Agent({
  keepAlive: true,
});

(async () => {
  await fetch('http://localhost:3000/testkeepalive?name=1', {
    agent: httpAgent,
  });
  await new Promise((res) => setTimeout(() => res(true), 2000));
  await fetch('http://localhost:3000/testkeepalive?name=2', {
    agent: httpAgent,
  });
  await new Promise((res) => setTimeout(() => res(true), 2000));
  await fetch('http://localhost:3000/testkeepalive?name=3', {
    agent: httpAgent,
  });
})();
```

The resuls make me think the keepAlive worked:

![Screenshot from 2023-07-06 18-18-00](https://github.com/nodejs/node/assets/50461902/1c300e7a-b26e-45ab-9246-d80345acbad9)
![image](https://github.com/nodejs/node/assets/50461902/e0e36e0d-b23d-4838-a07b-5639c9b6430b)

**Comment by bnoordhuis** [comment]
I think you're confusing HTTP Keep-Alive and TCP keepalive. Different layers of the stack.

**Reply by linhx** [reply]
I read the docs. and understand they are both for TCP.

Agent:

> keepAlive [<boolean>](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Boolean_type) Keep sockets around even when there are no outstanding requests, so they can be used for future requests without having to reestablish a TCP connection.

Server:

> similarly on what is done in [socket.setKeepAlive([enable][, initialDelay])][socket.setKeepAlive(enable, initialDelay)]

Could you please explain which one is HTTP Keep-Alive, which one is TCP keepalive? and what is the difference. Thank you.

**Reply by bnoordhuis** [reply]
> I read the docs. and understand they are both for TCP.

Not quite. There are three things getting conflated here:

1. The HTTP Keep-Alive header that communicates between server and client that it's okay to keep TCP connections open after the HTTP request completes,

2. The keepAlive option to http.Agent and http.Server that tells them to keep TCP connections open when there are no inflight HTTP requests or responses, and finally

3. TCP keepalive that instructs the OS to keep inactive TCP connections open for longer than it otherwise would, by sending out probe packets over the wire

socket.setKeepAlive() turns on the third one. The keepAlive option also turns on TCP keepalive but that's more of an implementation detail, its main purpose is to simply not close connections immediately.
